Panic: send on closed channel during shutdown

# Issue
Stopped Goshimmer with ctrl-c
Node Initiated shutdown but ran into panic error

# Info
Pi 4 Ubuntu aarm64
Goshimmer 0.1.0
statusscreen plugun disabled

# Log
`2020/02/04 05:50:20 [ INFO ] Autopeering: Discovered: 159.69.9.6:14625 / 64238185dd19a35e
2020/02/04 05:50:26 [ INFO ] Autopeering: Peering accepted: 159.69.9.6:14625 / 64238185dd19a35e
2020/02/04 05:50:30 [ INFO ] Autopeering: Discovered: 116.203.53.228:14626 / 50e18e3d8bd92716
2020/02/04 05:50:30 [ INFO ] Gossip: Connection to neighbor failed: 159.69.9.6:15625 / 64238185dd19a35e
2020/02/04 05:50:30 [ INFO ] Autopeering: Peering dropped: 64238185dd19a35e
2020/02/04 05:50:31 [ INFO ] Autopeering: Peering chosen: 116.203.53.228:14626 / 50e18e3d8bd92716
2020/02/04 05:50:32 [ INFO ] Gossip: Connection to neighbor failed: 116.203.53.228:14666 / 50e18e3d8bd92716
2020/02/04 05:50:32 [ INFO ] Autopeering: Peering dropped: 50e18e3d8bd92716
2020/02/04 05:50:40 [ INFO ] Autopeering: Discovered: 144.91.120.246:14626 / 0f7d30a14e377e06
2020/02/04 05:50:40 [ INFO ] Autopeering: Peering chosen: 144.91.120.246:14626 / 0f7d30a14e377e06
2020/02/04 05:50:41 [ INFO ] Gossip: Connection established     {"id": "0f7d30a14e377e06", "network": "tcp", "addr": "144.91.120.246:14666"}
2020/02/04 05:50:41 [ INFO ] Gossip: Neighbor added: 144.91.120.246:14666 / 0f7d30a14e377e06
^C2020/02/04 05:50:44 [ WARN ] Graceful Shutdown: Received shutdown request - waiting (max 10) to finish processing ...
2020/02/04 05:50:44 [ INFO ] SPA: Stopping SPA[WSSend] ...
2020/02/04 05:50:44 [ INFO ] SPA: Stopping SPA[WSSend] ... done
2020/02/04 05:50:44 [ INFO ] SPA: Stopping SPA[TxUpdater] ...
2020/02/04 05:50:44 [ INFO ] SPA: Stopping SPA[TxUpdater] ... done
2020/02/04 05:50:44 [ INFO ] WebAPI: Stopping Web Server ...
2020/02/04 05:50:44 [ INFO ] WebAPI: Stopping Web Server ... done
2020/02/04 05:50:44 [ INFO ] Gossip: Stopping Gossip ...
2020/02/04 05:50:44 [ INFO ] Gossip: Connection closed  {"id": "0f7d30a14e377e06", "network": "tcp", "addr": "144.91.120.246:14666"}
panic: send on closed channel

goroutine 111 [running]:
github.com/iotaledger/goshimmer/packages/gossip.(*Neighbor).Write(0x40004a5770, 0x401e41a960, 0x54, 0x60, 0x4000011e58, 0x1, 0x1)
        /opt/goshimmer/packages/gossip/neighbor.go:148 +0x54
github.com/iotaledger/goshimmer/packages/gossip.(*Manager).send(0x40003c43c0, 0x401e41a960, 0x54, 0x60, 0x0, 0x0, 0x0)
        /opt/goshimmer/packages/gossip/manager.go:176 +0x98
github.com/iotaledger/goshimmer/packages/gossip.(*Manager).RequestTransaction(0x40003c43c0, 0x401e41a720, 0x51, 0x51, 0x0, 0x0, 0x0)
        /opt/goshimmer/packages/gossip/manager.go:127 +0x178
github.com/iotaledger/goshimmer/plugins/gossip.requestTransaction(0x401e41a720, 0x51)
        /opt/goshimmer/plugins/gossip/gossip.go:125 +0x78
github.com/iotaledger/goshimmer/plugins/tangle.RequesterFunc.RequestTransaction(0xf19450, 0x401e41a720, 0x51)
        /opt/goshimmer/plugins/tangle/plugin.go:65 +0x38
github.com/iotaledger/goshimmer/plugins/tangle.requestTransaction(0x401e41a720, 0x51)
        /opt/goshimmer/plugins/tangle/solidifier.go:264 +0xf4
github.com/iotaledger/goshimmer/plugins/tangle.contains(0x401e41a720, 0x51, 0x51)
        /opt/goshimmer/plugins/tangle/solidifier.go:181 +0x90
github.com/iotaledger/goshimmer/plugins/tangle.isSolid(0x401e3dac00, 0x51, 0x51, 0x1, 0x1)
        /opt/goshimmer/plugins/tangle/solidifier.go:165 +0xd4
github.com/iotaledger/goshimmer/plugins/tangle.processNewTransaction(0x401e221360)
        /opt/goshimmer/plugins/tangle/solidifier.go:125 +0x200
github.com/iotaledger/goshimmer/plugins/tangle.processMetaTransaction(0x401e40e540)
        /opt/goshimmer/plugins/tangle/solidifier.go:88 +0x16c
github.com/iotaledger/goshimmer/plugins/tangle.configureSolidifier.func1(0x401e3f23f0, 0x1, 0x1, 0x401e251c20)
        /opt/goshimmer/plugins/tangle/solidifier.go:37 +0x4c
github.com/iotaledger/hive.go/workerpool.(*WorkerPool).startWorkers.func1(0x40004a47d0)
        /root/go/pkg/mod/github.com/iotaledger/hive.go@v0.0.0-20200121213505-28904d5f037c/workerpool/workerpool.go:159 +0x150
created by github.com/iotaledger/hive.go/workerpool.(*WorkerPool).startWorkers
        /root/go/pkg/mod/github.com/iotaledger/hive.go@v0.0.0-20200121213505-28904d5f037c/workerpool/workerpool.go:135 +0x5c
root@shimmy:/opt/goshimmer# ^C
root@shimmy:/opt/goshimmer# ^C
root@shimmy:/opt/goshimmer# `

--- a/packages/gossip/neighbor.go
+++ b/packages/gossip/neighbor.go
@@ -150,7 +150,7 @@ func (n *Neighbor) Write(b []byte) (int, error) {
 	case n.queue <- b:
 		return l, nil
 	case <-n.closing:
-		return 0, ErrNeighborQueueFull
+		return 0, ErrNeighborClosed
 	default:
 		return 0, ErrNeighborQueueFull
 	}

--- a/packages/gossip/neighbor.go
+++ b/packages/gossip/neighbor.go
@@ -16,8 +16,6 @@ import (
 var (
 	// ErrNeighborQueueFull is returned when the send queue is already full.
 	ErrNeighborQueueFull = errors.New("send queue is full")
-	// ErrNeighborClosed is returned when data was written to an already closed neighbor.
-	ErrNeighborClosed = errors.New("neighbor connection closed")
 )
 
 const (
@@ -87,7 +85,6 @@ func (n *Neighbor) disconnect() (err error) {
 	n.disconnectOnce.Do(func() {
 		close(n.closing)
 		err = n.BufferedConnection.Close()
-		close(n.queue)
 	})
 	return
 }
@@ -142,15 +139,13 @@ func (n *Neighbor) readLoop() {
 func (n *Neighbor) Write(b []byte) (int, error) {
 	l := len(b)
 	if l > maxPacketSize {
-		n.log.Errorw("message too large", "len", l, "max", maxPacketSize)
+		n.log.Panicw("message too large", "len", l, "max", maxPacketSize)
 	}
 
 	// add to queue
 	select {
 	case n.queue <- b:
 		return l, nil
-	case <-n.closing:
-		return 0, ErrNeighborClosed
 	default:
 		return 0, ErrNeighborQueueFull
 	}

--- a/packages/gossip/neighbor.go
+++ b/packages/gossip/neighbor.go
@@ -16,6 +16,8 @@ import (
 var (
 	// ErrNeighborQueueFull is returned when the send queue is already full.
 	ErrNeighborQueueFull = errors.New("send queue is full")
+	// ErrNeighborClosed is returned when data was written to an already closed neighbor.
+	ErrNeighborClosed = errors.New("neighbor connection closed")
 )
 
 const (
@@ -84,8 +86,8 @@ func (n *Neighbor) IsOutbound() bool {
 func (n *Neighbor) disconnect() (err error) {
 	n.disconnectOnce.Do(func() {
 		close(n.closing)
-		close(n.queue)
 		err = n.BufferedConnection.Close()
+		close(n.queue)
 	})
 	return
 }
@@ -147,6 +149,8 @@ func (n *Neighbor) Write(b []byte) (int, error) {
 	select {
 	case n.queue <- b:
 		return l, nil
+	case <-n.closing:
+		return 0, ErrNeighborQueueFull
 	default:
 		return 0, ErrNeighborQueueFull
 	}

--- a/packages/gossip/neighbor_test.go
+++ b/packages/gossip/neighbor_test.go
@@ -1,7 +1,6 @@
 package gossip
 
 import (
-	"errors"
 	"net"
 	"sync"
 	"sync/atomic"
@@ -34,18 +33,6 @@ func TestNeighborCloseTwice(t *testing.T) {
 	n.Listen()
 	require.NoError(t, n.Close())
 	require.NoError(t, n.Close())
-}
-
-func TestNeighborWriteToClosed(t *testing.T) {
-	a, _, teardown := newPipe()
-	defer teardown()
-
-	n := newTestNeighbor("A", a)
-	n.Listen()
-	require.NoError(t, n.Close())
-
-	_, err := n.Write(testData)
-	assert.True(t, errors.Is(err, ErrNeighborClosed), "unexpected error: %s", err)
 }
 
 func TestNeighborWrite(t *testing.T) {

--- a/packages/gossip/neighbor_test.go
+++ b/packages/gossip/neighbor_test.go
@@ -1,6 +1,7 @@
 package gossip
 
 import (
+	"errors"
 	"net"
 	"sync"
 	"sync/atomic"
@@ -43,9 +44,8 @@ func TestNeighborWriteToClosed(t *testing.T) {
 	n.Listen()
 	require.NoError(t, n.Close())
 
-	assert.Panics(t, func() {
-		_, _ = n.Write(testData)
-	})
+	_, err := n.Write(testData)
+	assert.True(t, errors.Is(err, ErrNeighborClosed), "unexpected error: %s", err)
 }
 
 func TestNeighborWrite(t *testing.T) {

